### PR TITLE
option: Do not permit setting kvstore type in CRD allocation mode

### LINKED ISSUE #11820: identity: deadlock with kvstore and identity-allocation-mode=crd
As discovered in https://github.com/cilium/cilium/issues/11690, there appears to be a deadlock when `identity-allocation-mode=crd` is used in combination with kvstore. Personally I don't know if this combination of options should be permitted, just reporting this as a follow-up, as right now it's possible for user to set it up this way and `cilium-agent` will lock-up (https://github.com/cilium/cilium/issues/11690#issuecomment-637044512).

### PR DESCRIPTION
**Pull request by vadorovsky**
CRD allocation mode should mean that kvstore is not used. Before this
change, user could set identity-allocation-mode=crd and kvstore options,
which resulted in trigerring the deamon to initialize a kvstore while
using CRD as the store, which resulted in a deadlock (#11690).

Fixes: 8822c1d650f0 ("k8s: Add CRD Identities as an identity allocator backend")
Fixes: #11820
Signed-off-by: Michal Rostecki <mrostecki@opensuse.org>

### CODE REVIEW

**Comment by mrostecki** [comment]
test-me-please

**Comment by mrostecki** [comment]
test-me-please

File: pkg/option/config.go [hunk 1]
```diff
@@ -2509,6 +2484,47 @@ func (c *DaemonConfig) populateHostServicesProtos() error {
 	return nil
 }
 
+func (c *DaemonConfig) populateIdentityAllocation() error {
+	c.DisableCiliumEndpointCRD = viper.GetBool(DisableCiliumEndpointCRDName)
```
Comment (joestringer) [inline pkg/option/config.go:2488]: How does CiliumEndpoint CRD relate to identities? ciliumidentities sure, but ciliumendpoint I'm not sure.

File: pkg/option/config_test.go [hunk 2]
```diff
@@ -700,6 +700,151 @@ func Test_populateNodePortRange(t *testing.T) {
...
+				wantErr:                      false,
+			},
+			preTestRun: func() {
+				viper.Reset()
+				viper.Set(IdentityAllocationMode, IdentityAllocationModeCRD)
+				viper.Set(KVStore, "etcd")
+			},
+		},
+		{
+			name: "crd mode provided, endpoint crd enabled, should enable endpoint crd",
```
Comment (joestringer) [inline pkg/option/config_test.go:758]: nit:

```suggestion
			name: "crd mode provided, endpoint crd disabled, should enable endpoint crd",
```

File: pkg/option/config_test.go [hunk 3]
```diff
@@ -700,6 +700,151 @@ func Test_populateNodePortRange(t *testing.T) {
...
+				wantIdentityAllocationMode:   IdentityAllocationModeCRD,
+				wantKVStore:                  "",
+				wantDisableCiliumEndpointCRD: false,
+				wantErr:                      false,
+			},
+			preTestRun: func() {
+				viper.Reset()
+				viper.Set(IdentityAllocationMode, IdentityAllocationModeKVstore)
+				viper.Set(KVStore, "")
+			},
```
Comment (joestringer) [inline pkg/option/config_test.go:783]: Hmm, I kinda wonder if someone explicitly passes kvstore mode but fails to properly configure it whether it should actually error out..?

File: pkg/option/config.go [hunk 4]
```diff
@@ -2509,6 +2484,47 @@ func (c *DaemonConfig) populateHostServicesProtos() error {
...
+
+	if c.KVStore == "" {
+		if c.IdentityAllocationMode != IdentityAllocationModeCRD {
+			log.Warningf("Running Cilium with %q=%q requires identity allocation via CRDs. Changing %s to %q", KVStore, c.KVStore, IdentityAllocationMode, IdentityAllocationModeCRD)
+			c.IdentityAllocationMode = IdentityAllocationModeCRD
+		}
+		if c.DisableCiliumEndpointCRD {
+			log.Warningf("Running Cilium with %q=%q requires endpoint CRDs. Changing %s to %t", KVStore, c.KVStore, DisableCiliumEndpointCRDName, false)
+			c.DisableCiliumEndpointCRD = false
+		}
```
Comment (joestringer) [inline pkg/option/config.go:2522]: @aanm do you know if this is still true? We have ciliumidentities CRD now, is this left over or we still have a dependency on ciliumendpoint CRD as well?

**Review by joestringer — CHANGES_REQUESTED** [review]
Couple of minor nits in the testsuite side.

I raised some question in the main logic but it LGTM.

**Comment by mrostecki** [comment]
I'm closing this PR, since my assumption about kvstore and CRD being mutually exclusive was wrong. Sorry!

**Pull request closed without merging**